app/assets/javascripts/discourse/app/lib/plugin-api.js

```diff
@@ -1926,6 +1927,37 @@ class PluginApi {
   registerUserMenuTab(func) {
     registerUserMenuTab(func);
   }
+
+  /**
+   * EXPERIMENTAL. Do not use.
+   * Apply transformation using a callback on a list of model instances of a
+   * specific type. Currently, this API only works on lists rendered in the
+   * user menu such as notifications, bookmarks and topics (i.e. messages), but
+   * it may be extended to other lists in other parts of the app.
+   *
+   * You can pass an `async` callback to this API and it'll be `await`ed and
+   * block rendering until the callback finishes executing.
+   *
+   * ```
+   * api.registerModelTransformer("topic", async (topics) => {
+   *   for (const topic of topics) {
+   *     const decryptedTitle = await decryptTitle(topic.encrypted_title);
+   *     if (decryptedTitle) {
+   *       topic.fancy_title = decryptedTitle;
+   *     }
+   *   }
+   * });
+   * ```
+   *
+   * @callback registerModelTransformerCallback
+   * @param {Object[]} A list of model instances
+   *
+   * @param {string} modelName - Model type on which transformation should be applied. Currently the only valid type is "topic".
+   * @param {registerModelTransformerCallback} transformer - Callback function that receives a list of model objects of the specified type and applies transformation on them.
+   */
```

Comment:
I think it'll be better if we can enforce that the `registerModelTransformerCallback` must return a Promise so that we can be sure that we're working with and awaiting a promise instead of some random data type.

Reply:
What if the transformation can be done synchronously? In that case, it would make a lot of sense for someone to pass a regular function as a callback.

Reply:
Hmm I think most of transformation that plugins apply don't need to do anything asynchronous so I don't think it'd be fair to require plugins to return `Promise` when they're not doing anything asynchronous. Even if a plugin decides to be weird and return something random in the callback, nothing would break:

```js
syncFunc = () => 4;
asyncFunc = async () => {
    return new Promise((res) => {
        setTimeout(() => res(4), 10)
    })
}

console.log(await syncFunc()) // prints 4
console.log(await asyncFunc()) // also prints 4
```

Reply:
Ahh icic we can await on any data type. When reading the docs it seems like await has to be used on a promise.